strawberryfields/configuration.py

```diff
@@ -15,138 +15,229 @@
+        filename (str): the configuration file to look for
+
+    Returns:
+         dict[str, dict[str, Union[str, bool, int]]] or None: the
+             configuration object that was loaded
+    """
+
+    # Search the current directory, the directory under environment
+    # variable SF_CONF, and default user config directory, in that order.
+    current_dir = os.getcwd()
+    sf_env_config_dir = os.environ.get("SF_CONF", "")
+    sf_user_config_dir = user_config_dir("strawberryfields", "Xanadu")
+
+    directories = [current_dir, sf_env_config_dir, sf_user_config_dir]
+    for directory in directories:
+        filepath = os.path.join(directory, filename)
+        try:
+            parsed_config = load_config_file(filepath)
+            break
+        except FileNotFoundError:
+            parsed_config, filepath = None, None
+
+    return parsed_config, filepath
+
+def load_config_file(filepath):
+    """Load a configuration object from a TOML formatted file.
+
+    Args:
+        filepath (str): path to the configuration file
+
+    Returns:
+         dict[str, dict[str, Union[str, bool, int]]]: the configuration
+            object that was loaded
+    """
+    with open(filepath, "r") as f:
+        config_from_file = toml.load(f)
+    return config_from_file
+
+def update_with_other_config(config, other_config):
+    """Updates the current configuration object with another one.
+
+    Args:
+        config (dict[str, dict[str, Union[str, bool, int]]]): the
+            configuration to be updated
+        other_config (dict[str, dict[str, Union[str, bool, int]]]): the
+            configuration used for updating
+
+    Returns:
+        dict[str, dict[str, Union[str, bool, int]]]): the updated
+            configuration
+    """
+    # Here an example for section is API
+    for section, sectionconfig in config.items():
+        for key in sectionconfig:
+            if key in other_config[section]:
+                config[section][key] = other_config[section][key]
+
+def update_from_environment_variables(config):
+    """Updates the current configuration object from data stored in environment
+    variables.
+
+    .. note::
+
+        The following environment variables are checked:
+
+        * SF_API_AUTHENTICATION_TOKEN
+        * SF_API_HOSTNAME
+        * SF_API_USE_SSL
+        * SF_API_DEBUG
+        * SF_API_PORT
+
+    Args:
+        config (dict[str, dict[str, Union[str, bool, int]]]): the
+            configuration to be updated
+    Returns:
+        dict[str, dict[str, Union[str, bool, int]]]): the updated
+        configuration
+    """
+    for section, sectionconfig in config.items():
+        env_prefix = "SF_{}_".format(section.upper())
+        for key in sectionconfig:
+            env = env_prefix + key.upper()
+            if env in os.environ:
+                config[section][key] = parse_environment_variable(key, os.environ[env])
+
+
+BOOLEAN_KEYS = ("debug", "use_ssl")
+INTEGER_KEYS = ("port")
```

Comment:
Since these are defined as module level variables, I would suggest:

1. Moving them to the top of the file

2. Including the default values as well.

These could then be programmatically accessed from `create_config`. This is nice from a readability point of view --- if the default keys/types of the keys ever need to be updated, there is a 'default_config' dict at the top of the file that contains the sections, keys, types, and default values.

Reply:
I've moved this to the top now :) As for a default config, it is accessible from `DEFAULT_CONFIG` which calls on `create_config` with no parameters. Would you prefer to have it such that it is explicitly included as well?

Reply:
No I think that's a good way of doing it!